File: app/controllers/orders_controller.rb
class OrdersController < ApplicationController
  ORDER_FIELDS = %i[order_id status purchase_at estimated_delivery_at delivered_customer_at].freeze
  SEARCH_STRATEGIES = %w[n_plus_one lazy_loading includes preload eager_load joins raw_sql].freeze
  DEFAULT_SEARCH_STRATEGY = "n_plus_one"
  ASSOCIATION_TREE = [
    :customer,
    :order_payments,
    :order_reviews,
    { order_items: %i[product seller] }
  ].freeze

  # PostgreSQL-specific on purpose: this study strategy builds the complete public
  # graph in one SELECT with JSONB aggregates instead of instantiating AR models.
  RAW_SQL_QUERY = <<~SQL.freeze
    SELECT
      jsonb_build_object(
        'order_id', selected_orders.order_id,
        'status', selected_orders.status,
        'purchase_at', to_char(selected_orders.purchase_at, 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"'),
        'estimated_delivery_at', to_char(selected_orders.estimated_delivery_at, 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"'),
        'delivered_customer_at', CASE
          WHEN selected_orders.delivered_customer_at IS NULL THEN NULL
          ELSE to_char(selected_orders.delivered_customer_at, 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"')
        END,
        'customer', CASE
          WHEN customers.id IS NULL THEN NULL
          ELSE jsonb_build_object(
            'customer_id', customers.customer_id,
            'customer_unique_id', customers.customer_unique_id,
            'city', customers.city,
            'state', customers.state
          )
        END,
        'items', items.data,
        'payments', payments.data,
        'reviews', reviews.data
      ) AS order_data
    FROM (
      SELECT
        id,
        customer_id,
        order_id,
        status,
        purchase_at,
        estimated_delivery_at,
        delivered_customer_at
      FROM orders
      ORDER BY purchase_at DESC, order_id ASC
      LIMIT 250
    ) AS selected_orders
    LEFT JOIN customers ON customers.id = selected_orders.customer_id
    LEFT JOIN LATERAL (
      SELECT COALESCE(
        jsonb_agg(
          jsonb_build_object(
            'order_item_id', order_items.order_item_id,
            'product_id', products.product_id,
            'seller_id', sellers.seller_id,
            'price', to_char(order_items.price, 'FM9999999990.00'),
            'freight_value', to_char(order_items.freight_value, 'FM9999999990.00')
          ) ORDER BY order_items.order_item_id ASC
        ),
        '[]'::jsonb
      ) AS data
      FROM order_items
      LEFT JOIN products ON products.id = order_items.product_id
      LEFT JOIN sellers ON sellers.id = order_items.seller_id
      WHERE order_items.order_id = selected_orders.id
    ) AS items ON TRUE
    LEFT JOIN LATERAL (
      SELECT COALESCE(
        jsonb_agg(
          jsonb_build_object(
            'payment_sequential', order_payments.payment_sequential,
            'payment_type', order_payments.payment_type,
            'payment_installments', order_payments.payment_installments,
            'payment_value', to_char(order_payments.payment_value, 'FM9999999990.00')
          ) ORDER BY order_payments.payment_sequential ASC
        ),
        '[]'::jsonb
      ) AS data
      FROM order_payments
      WHERE order_payments.order_id = selected_orders.id
    ) AS payments ON TRUE
    LEFT JOIN LATERAL (
      SELECT COALESCE(
        jsonb_agg(
          jsonb_build_object(
            'review_id', order_reviews.review_id,
            'score', order_reviews.score,
            'comment_title', order_reviews.comment_title,
            'comment_message', order_reviews.comment_message,
            'creation_at', to_char(order_reviews.creation_at, 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"'),
            'answer_at', to_char(order_reviews.answer_at, 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"')
          ) ORDER BY order_reviews.creation_at ASC, order_reviews.review_id ASC, order_reviews.id ASC
        ),
        '[]'::jsonb
      ) AS data
      FROM order_reviews
      WHERE order_reviews.order_id = selected_orders.id
    ) AS reviews ON TRUE
    ORDER BY selected_orders.purchase_at DESC, selected_orders.order_id ASC
  SQL
  private_constant :RAW_SQL_QUERY

  def index
    respond_to do |format|
      format.html
      format.json do
        requested_strategy = params[:strategy].presence || DEFAULT_SEARCH_STRATEGY

        unless SEARCH_STRATEGIES.include?(requested_strategy)
          render json: { error: "invalid_strategy" }, status: :unprocessable_content
          next
        end

        serialized_orders = measure_order_queries(
          requested_strategy: requested_strategy,
          effective_strategy: requested_strategy
        ) do
          case requested_strategy
          when "n_plus_one" then n_plus_one_query
          when "lazy_loading" then lazy_loading_query
          when "includes" then includes_query
          when "preload" then preload_query
          when "eager_load" then eager_load_query
          when "joins" then joins_query
          when "raw_sql" then raw_sql_query
          end
        end

        render json: {
          strategy: {
            requested: requested_strategy,
            effective: requested_strategy,
            implemented: true
          },
          orders: serialized_orders
        }
      end
    end
  end

  private

  def n_plus_one_query
    serialize_orders(base_order_scope)
  end

  def lazy_loading_query
    # O acesso às associações no serializer é lazy e reproduz o N+1 neste exemplo.
    serialize_orders(base_order_scope)
  end

  def includes_query
    serialize_orders(base_order_scope.includes(ASSOCIATION_TREE))
  end

  def preload_query
    serialize_orders(base_order_scope.preload(ASSOCIATION_TREE))
  end

  def eager_load_query
    serialize_orders(base_order_scope.eager_load(ASSOCIATION_TREE).reorder(purchase_at: :desc, order_id: :asc))
  end

  def joins_query
    serialize_orders(base_order_scope.joins(:customer))
  end

  def raw_sql_query
    ActiveRecord::Base.connection.select_all(RAW_SQL_QUERY).map do |row|
      order_data = row.fetch("order_data")
      order_data = JSON.parse(order_data) if order_data.is_a?(String)
      order_data.deep_stringify_keys
    end
  end

  def base_order_scope
    Order.order(purchase_at: :desc, order_id: :asc).limit(250)
  end

  def serialize_orders(orders)
    orders.map do |order|
      customer = order.customer
      items = order.order_items.sort_by(&:order_item_id).map do |item|
        product = item.product
        seller = item.seller

        {
          order_item_id: item.order_item_id,
          product_id: product&.product_id,
          seller_id: seller&.seller_id,
          price: format("%.2f", item.price),
          freight_value: format("%.2f", item.freight_value)
        }
      end
      payments = order.order_payments.sort_by(&:payment_sequential).map do |payment|
        {
          payment_sequential: payment.payment_sequential,
          payment_type: payment.payment_type,
          payment_installments: payment.payment_installments,
          payment_value: format("%.2f", payment.payment_value)
        }
      end
      reviews = order.order_reviews.sort_by { |review| [ review.creation_at, review.review_id, review.id ] }.map do |review|
        {
          review_id: review.review_id,
          score: review.score,
          comment_title: review.comment_title,
          comment_message: review.comment_message,
          creation_at: review.creation_at,
          answer_at: review.answer_at
        }
      end

      order.as_json(only: ORDER_FIELDS).merge(
        "customer" => customer&.as_json(only: %i[customer_id customer_unique_id city state]),
        "items" => items,
        "payments" => payments,
        "reviews" => reviews
      )
    end
  end

  def measure_order_queries(requested_strategy:, effective_strategy:)
    origin_thread = Thread.current
    origin_fiber = Fiber.current
    query_count = 0
    db_duration_ms = 0.0
    total_started_at = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    callback = lambda do |_name, started_at, finished_at, _event_id, payload|
      next unless Thread.current.equal?(origin_thread) && Fiber.current.equal?(origin_fiber)
      next if payload[:cached]
      next if payload[:name].to_s.match?(/SCHEMA|SchemaMigration/)
      next if payload[:name] == "TRANSACTION"

      normalized_sql = payload[:sql].to_s.gsub(/\s+/, " ").strip
      next unless normalized_sql.match?(/\ASELECT\b/i)

      duration_ms = (finished_at - started_at) * 1_000
      query_count += 1
      db_duration_ms += duration_ms
      Rails.logger.info(
        "[ORDERS_QUERY_STUDY] query=#{query_count} " \
          "requested_strategy=#{requested_strategy} effective_strategy=#{effective_strategy} " \
          "duration_ms=#{format('%.3f', duration_ms)} sql=#{normalized_sql.inspect}"
      )
    end

    ActiveSupport::Notifications.subscribed(callback, "sql.active_record", monotonic: true) do
      yield
    end
  ensure
    if total_started_at
      total_duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - total_started_at) * 1_000
      Rails.logger.info(
        "[ORDERS_QUERY_STUDY] summary=true status=#{$! ? 'error' : 'ok'} " \
          "requested_strategy=#{requested_strategy} effective_strategy=#{effective_strategy} " \
          "query_count=#{query_count} db_duration_ms=#{format('%.3f', db_duration_ms)} " \
          "total_duration_ms=#{format('%.3f', total_duration_ms)}"
      )
    end
  end
end

File: app/javascript/controllers/orders_controller.js
import { Controller } from "@hotwired/stimulus"

const STRATEGIES = Object.freeze({
  n_plus_one: {
    label: "N+1 intencional",
    help: "Carrega as associações uma a uma de propósito. Compare as consultas e os tempos no terminal."
  },
  lazy_loading: {
    label: "Lazy loading",
    help: "Carrega cada associação quando ela é acessada; neste exemplo, pode repetir o mesmo padrão do N+1. Compare os logs no terminal."
  },
  includes: {
    label: "Eager loading com includes",
    help: "Deixa o Rails escolher como pré-carregar as associações. Compare as consultas e os tempos no terminal."
  },
  preload: {
    label: "Eager loading com preload",
    help: "Pré-carrega as associações em consultas separadas e em lote. Compare as consultas e os tempos no terminal."
  },
  eager_load: {
    label: "Eager loading com eager_load",
    help: "Pré-carrega as associações com LEFT OUTER JOIN. Compare as consultas e os tempos no terminal."
  },
  joins: {
    label: "JOIN de cliente via joins",
    help: "Executa joins(:customer): aplica INNER JOIN apenas ao cliente e não pré-carrega nenhuma associação."
  },
  raw_sql: {
    label: "SQL puro (PostgreSQL)",
    help: "Executa uma única consulta SQL do PostgreSQL, com agregações JSONB e LATERAL, sem carregar associações pelo Active Record. Não é portátil para outros bancos."
  }
})

export default class extends Controller {
  static targets = ["button", "error", "results", "status", "strategy", "strategyHelp"]
  static values = { url: String }

  connect() {
    this.updateStrategyHelp()
  }

  async load() {
    const selection = this.selectedStrategy()
    this.startLoading(selection)

    try {
      if (!selection) throw new Error("Invalid selected strategy")

      const url = new URL(this.urlValue, window.location.origin)
      url.searchParams.set("strategy", selection.requested)

      const response = await fetch(url, {
        headers: { Accept: "application/json" },
        credentials: "same-origin"
      })

      if (!response.ok) throw new Error(`Request failed with status ${response.status}`)

      const payload = await response.json()
      this.validatePayload(payload, selection)

      this.renderOrders(payload.orders, selection)
    } catch (_error) {
      this.showError(selection)
    } finally {
      this.finishLoading()
    }
  }

  selectedStrategy() {
    const requested = this.strategyTarget.value
    const strategy = STRATEGIES[requested]

    if (!strategy) return null

    return { requested, ...strategy }
  }

  updateStrategyHelp() {
    const selection = this.selectedStrategy()
    this.strategyHelpTarget.textContent = selection?.help || ""
  }

  validatePayload(payload, selection) {
    if (!payload || typeof payload !== "object" || Array.isArray(payload)) throw new Error("Invalid orders response")
    if (!Array.isArray(payload.orders)) throw new Error("Invalid orders response")

    const strategy = payload.strategy
    if (!strategy || typeof strategy !== "object" || Array.isArray(strategy)) throw new Error("Invalid strategy response")
    if (strategy.requested !== selection.requested) throw new Error("Mismatched strategy response")
    if (strategy.effective !== selection.requested) throw new Error("Invalid effective strategy")
    if (strategy.implemented !== true) throw new Error("Invalid implementation status")
  }

  startLoading(selection) {
    this.buttonTarget.disabled = true
    this.strategyTarget.disabled = true
    this.buttonTarget.textContent = "Buscando…"
    this.resultsTarget.replaceChildren()
    this.resultsTarget.setAttribute("aria-busy", "true")
    this.statusTarget.textContent = selection ? `Buscando até 250 pedidos com ${selection.label}…` : ""
    this.errorTarget.hidden = true
    this.errorTarget.textContent = ""
  }

  finishLoading() {
    this.buttonTarget.disabled = false
    this.strategyTarget.disabled = false
    this.buttonTarget.textContent = "Buscar pedidos"
    this.resultsTarget.setAttribute("aria-busy", "false")
  }

  renderOrders(orders, strategy) {
    if (orders.length === 0) {
      this.statusTarget.textContent = `Nenhum pedido encontrado com ${strategy.label}.`
      return
    }

    const list = document.createElement("ul")
    list.className = "orders-list"

    orders.forEach((order) => list.append(this.buildOrderItem(order)))
    this.resultsTarget.append(list)

    const noun = orders.length === 1 ? "pedido encontrado" : "pedidos encontrados"
    this.statusTarget.textContent = `${orders.length} ${noun} com ${strategy.label}.`
  }

  buildOrderItem(order) {
    const item = document.createElement("li")
    item.className = "order-card"
    const safeOrder = order && typeof order === "object" ? order : {}
    const heading = document.createElement("h2")
    const baseDetails = document.createElement("dl")
    const relatedDetails = document.createElement("details")
    const summary = document.createElement("summary")
    const relatedContent = document.createElement("div")
    const items = this.collection(safeOrder.items)
    const payments = this.collection(safeOrder.payments)
    const reviews = this.collection(safeOrder.reviews)

    heading.textContent = `Pedido ${this.displayValue(safeOrder.order_id)}`
    this.appendDetail(baseDetails, "Código do pedido", this.displayValue(safeOrder.order_id))
    this.appendDetail(baseDetails, "Status", this.displayValue(safeOrder.status))
    this.appendDetail(baseDetails, "Comprado em", this.formatDate(safeOrder.purchase_at))
    this.appendDetail(baseDetails, "Entrega estimada", this.formatDate(safeOrder.estimated_delivery_at))
    this.appendDetail(baseDetails, "Entregue em", this.formatDate(safeOrder.delivered_customer_at))

    relatedDetails.className = "order-related"
    summary.textContent = `Ver dados relacionados — ${this.quantity(items.length, "item", "itens")}, ${this.quantity(payments.length, "pagamento", "pagamentos")}, ${this.quantity(reviews.length, "avaliação", "avaliações")}`
    relatedContent.className = "order-related-content"
    relatedContent.append(
      this.buildCustomerSection(safeOrder.customer),
      this.buildItemsSection(items),
      this.buildPaymentsSection(payments),
      this.buildReviewsSection(reviews)
    )
    relatedDetails.append(summary, relatedContent)

    item.append(heading, baseDetails, relatedDetails)
    return item
  }

  buildCustomerSection(customer) {
    const section = this.section("Cliente")

    if (!customer || typeof customer !== "object" || Array.isArray(customer)) {
      section.append(this.emptyMessage("Cliente não informado."))
      return section
    }

    const details = document.createElement("dl")
    this.appendDetail(details, "ID do cliente", this.displayValue(customer.customer_id))
    this.appendDetail(details, "ID único do cliente", this.displayValue(customer.customer_unique_id))
    this.appendDetail(details, "Cidade", this.displayValue(customer.city))
    this.appendDetail(details, "Estado", this.displayValue(customer.state))
    section.append(details)
    return section
  }

  buildItemsSection(items) {
    const section = this.section("Itens")

    if (items.length === 0) {
      section.append(this.emptyMessage("Nenhum item relacionado."))
      return section
    }

    section.append(this.relatedList(items, (item) => {
      const details = document.createElement("dl")
      this.appendDetail(details, "Número do item", this.displayValue(item.order_item_id))
      this.appendDetail(details, "ID do produto", this.displayValue(item.product_id))
      this.appendDetail(details, "ID do vendedor", this.displayValue(item.seller_id))
      this.appendDetail(details, "Preço", this.formatMoney(item.price))
      this.appendDetail(details, "Frete", this.formatMoney(item.freight_value))
      return details
    }))
    return section
  }

  buildPaymentsSection(payments) {
    const section = this.section("Pagamentos")

    if (payments.length === 0) {
      section.append(this.emptyMessage("Nenhum pagamento relacionado."))
      return section
    }

    section.append(this.relatedList(payments, (payment) => {
      const details = document.createElement("dl")
      this.appendDetail(details, "Sequência", this.displayValue(payment.payment_sequential))
      this.appendDetail(details, "Tipo", this.displayValue(payment.payment_type))
      this.appendDetail(details, "Parcelas", this.displayValue(payment.payment_installments))
      this.appendDetail(details, "Valor", this.formatMoney(payment.payment_value))
      return details
    }))
    return section
  }

  buildReviewsSection(reviews) {
    const section = this.section("Avaliações")

    if (reviews.length === 0) {
      section.append(this.emptyMessage("Nenhuma avaliação relacionada."))
      return section
    }

    section.append(this.relatedList(reviews, (review) => {
      const details = document.createElement("dl")
      this.appendDetail(details, "ID da avaliação", this.displayValue(review.review_id))
      this.appendDetail(details, "Nota", this.displayValue(review.score))
      this.appendDetail(details, "Título", this.displayValue(review.comment_title))
      this.appendDetail(details, "Comentário", this.displayValue(review.comment_message))
      this.appendDetail(details, "Criada em", this.formatDate(review.creation_at))
      this.appendDetail(details, "Respondida em", this.formatDate(review.answer_at))
      return details
    }))
    return section
  }

  section(title) {
    const section = document.createElement("section")
    const heading = document.createElement("h3")

    heading.textContent = title
    section.append(heading)
    return section
  }

  relatedList(records, buildContent) {
    const list = document.createElement("ul")
    list.className = "order-related-list"

    records.forEach((record) => {
      const item = document.createElement("li")
      const safeRecord = record && typeof record === "object" ? record : {}
      item.append(buildContent(safeRecord))
      list.append(item)
    })
    return list
  }

  emptyMessage(message) {
    const paragraph = document.createElement("p")
    paragraph.className = "order-related-empty"
    paragraph.textContent = message
    return paragraph
  }

  appendDetail(details, label, value) {
    const group = document.createElement("div")
    const term = document.createElement("dt")
    const description = document.createElement("dd")

    term.textContent = label
    description.textContent = value
    group.append(term, description)
    details.append(group)
  }

  displayValue(value) {
    if (typeof value === "string") return value.trim() ? value : "Não informado"
    if (typeof value === "number" && Number.isFinite(value)) return String(value)

    return "Não informado"
  }

  collection(value) {
    return Array.isArray(value) ? value : []
  }

  quantity(count, singular, plural) {
    return `${count} ${count === 1 ? singular : plural}`
  }

  formatMoney(value) {
    if ((typeof value !== "string" && typeof value !== "number") || value === "") return "Não informado"

    const amount = Number(value)
    if (!Number.isFinite(amount)) return "Não informado"

    return new Intl.NumberFormat("pt-BR", {
      style: "currency",
      currency: "BRL",
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    }).format(amount)
  }

  formatDate(value) {
    if (typeof value !== "string" || !value) return "Não informado"

    const date = new Date(value)
    if (Number.isNaN(date.getTime())) return "Não informado"

    return new Intl.DateTimeFormat("pt-BR", {
      dateStyle: "short",
      timeStyle: "short"
    }).format(date)
  }

  showError(strategy) {
    this.statusTarget.textContent = ""
    const label = strategy?.label || "a estratégia selecionada"
    this.errorTarget.textContent = `Não foi possível carregar os pedidos com ${label}. Tente novamente.`
    this.errorTarget.hidden = false
  }
}
